src/ppx/xr_component.cpp

```diff
@@ -0,0 +1,395 @@
+    XrViewLocateInfo viewLocateInfo = {
+        .type                  = XR_TYPE_VIEW_LOCATE_INFO,
+        .viewConfigurationType = XR_VIEW_CONFIGURATION_TYPE_PRIMARY_STEREO,
+        .displayTime           = mFrameState.predictedDisplayTime,
+        .space                 = mRefSpace,
+    };
+
+    XrViewState viewState = {
+        .type = XR_TYPE_VIEW_STATE,
+    };
+    uint32_t viewCount = 0;
+    VERIFY_XR_CALL(xrLocateViews(mSession, &viewLocateInfo, &viewState, (uint32_t)mViews.size(), &viewCount, mViews.data()));
+
+    if ((viewState.viewStateFlags & XR_VIEW_STATE_POSITION_VALID_BIT) == 0 ||
+        (viewState.viewStateFlags & XR_VIEW_STATE_ORIENTATION_VALID_BIT) == 0) {
+        mShouldRender = false;
+    }
+
+    mCompositionLayerProjectionViews.resize(viewCount);
+    PPX_ASSERT_MSG(swapchains.size() == viewCount, "Swapchain needs to match number of views!");
+    for (uint32_t i = 0; i < viewCount; ++i) {
+        mCompositionLayerProjectionViews[i]                           = {XR_TYPE_COMPOSITION_LAYER_PROJECTION_VIEW};
+        mCompositionLayerProjectionViews[i].pose                      = mViews[i].pose;
+        mCompositionLayerProjectionViews[i].fov                       = mViews[i].fov;
+        mCompositionLayerProjectionViews[i].subImage.swapchain        = swapchains[i]->GetXrSwapchain();
+        mCompositionLayerProjectionViews[i].subImage.imageRect.offset = {0, 0};
+        mCompositionLayerProjectionViews[i].subImage.imageRect.extent = {static_cast<int>(GetWidth()), static_cast<int>(GetHeight())};
+    }
+
+    // Begin frame
+    XrFrameBeginInfo frameBeginInfo = {
+        .type = XR_TYPE_FRAME_BEGIN_INFO,
+    };
+
+    XrResult result = xrBeginFrame(mSession, &frameBeginInfo);
+    if (result != XR_SUCCESS) {
+        switch (result) {
+            case XR_SESSION_LOSS_PENDING:
+            case XR_FRAME_DISCARDED:
+                mShouldRender = false;
+                break;
+
+            default:
+                PPX_ASSERT_MSG(false, "xrBeginFrame failed!");
+                break;
+        }
+    }
+}
+
+void XrComponent::EndFrame()
+{
+    mCompositionLayerProjection.space     = mRefSpace;
+    mCompositionLayerProjection.viewCount = static_cast<uint32_t>(GetViewCount());
+    mCompositionLayerProjection.views     = mCompositionLayerProjectionViews.data();
+    XrCompositionLayerBaseHeader* layer   = mShouldRender ? reinterpret_cast<XrCompositionLayerBaseHeader*>(&mCompositionLayerProjection) : nullptr;
+
+    XrFrameEndInfo frameEndInfo = {
+        .type                 = XR_TYPE_FRAME_END_INFO,
+        .displayTime          = mFrameState.predictedDisplayTime,
+        .environmentBlendMode = mBlend,
+        .layerCount           = static_cast<uint32_t>(mShouldRender ? 1 : 0),
+        .layers               = &layer,
+    };
+
+    VERIFY_XR_CALL(xrEndFrame(mSession, &frameEndInfo));
+}
+
+glm::mat4 XrComponent::GetCurrentViewViewMatrix() const
```

Comment:
typo in function name

Reply:
lol, it is not, because it is "Get Current View's View Matrix". Agree that it is a bit confusing. Let me know if you have a better name option, and i am happy to change this :)

Reply:
maybe it is better to call it GetViewMatrixOfCurrentView?

Reply:
Yes, that sounds better!

Reply:
Maybe `ForCurrentView` instead of `OfCurrentView`

Reply:
fixed